src/main/java/org/xbill/DNS/lookup/LookupSession.java

```diff
@@ -0,0 +1,299 @@
+   * this is a bit odd, but they are inherited from Lookup.java. Note that the stream returned is
+   * never empty, as it will at the very least always contain name.
+   */
+  Stream<Name> expandName(Name name) throws NameTooLongException {
+    if (name.isAbsolute()) {
+      return Stream.of(name);
+    }
+    if (name.length() == 255) {
+      // handle this special case where the name is relative but can not be made absolute as it
+      // is at it's maximum length and nothing can be appended.
+      throw new NameTooLongException("Name is too long");
+    }
+    Stream<Name> fromSearchPath =
+        Stream.concat(
+            searchPath.stream()
+                .map(searchSuffix -> safeConcat(name, searchSuffix))
+                .filter(Objects::nonNull),
+            Stream.of(safeConcat(name, Name.root)));
+
+    if (name.labels() > ndots) {
+      return Stream.concat(Stream.of(safeConcat(name, Name.root)), fromSearchPath);
+    }
+    return fromSearchPath;
+  }
+
+  private static Name safeConcat(Name name, Name suffix) {
+    try {
+      return Name.concatenate(name, suffix);
+    } catch (NameTooLongException e) {
+      return null;
+    }
+  }
+
+  private void lookupUntilSuccess(
+      Iterator<Name> names, int type, int dclass, CompletableFuture<LookupResult> future) {
+
+    Record query = Record.newRecord(names.next(), type, dclass);
+    lookupWithCache(query, null)
+        .thenCompose(answer -> resolveRedirects(answer, query))
+        .whenComplete(
+            (result, ex) -> {
+              if (ex != null) {
+                future.completeExceptionally(ex);
+              } else if (!result.getRecords().isEmpty() || !names.hasNext()) {
+                future.complete(result);
+              } else {
+                lookupUntilSuccess(names, type, dclass, future);
+              }
+            });
+  }
+
+  private CompletionStage<LookupResult> lookupWithCache(Record queryRecord, List<Name> aliases) {
+
+    Supplier<CompletionStage<LookupResult>> lookupWithResolver =
+        () ->
+            resolver
+                .sendAsync(Message.newQuery(queryRecord))
+                .thenApply(this::maybeAddToCache)
+                .thenApply(answer -> dealWithEmptyAnswer(answer, aliases));
+
+    return Optional.ofNullable(caches.get(queryRecord.getDClass()))
+        .map(c -> c.lookupRecords(queryRecord.getName(), queryRecord.getType(), Credibility.NORMAL))
+        .map(this::setResponseToMessageFuture)
+        .orElseGet(lookupWithResolver);
+  }
+
+  private Message maybeAddToCache(Message message) {
+    Optional.ofNullable(caches.get(message.getQuestion().getDClass()))
+        .ifPresent(cache -> cache.addMessage(message));
+    return message;
+  }
+
+  private CompletionStage<LookupResult> setResponseToMessageFuture(SetResponse setResponse) {
+    if (setResponse.isNXDOMAIN()) {
+      return completedFuture(new LookupResult(emptyList(), AdditionalDetail.NXDOMAIN, null));
```

Comment:
I'm unused to static imports except for assertions and (sometimes) constants

Reply:
While this particular code has changed, for the general case I am reading your statement that you are unused to the practice in the context of a change request as that you would rather me not doing that. While I could definitely make a compromise and remove the static imports, I think that it improves readability. Please tell me more about what you think, why and how strong this opinion is :)

Reply:
I think it's confusing (with local method names). I'd refer to some guidelines, but I don't have any at hand. If you're used to it from previous work, then I guess I could live with it. I've just never seen static imports for methods used outside of test code.